p2p/protocol.go

```diff
@@ -43,21 +43,21 @@ type Protocol struct {
 	// The peer connection is closed when Start returns. It should return
 	// any protocol-level error (such as an I/O error) that is
 	// encountered.
-	Run func(peer *Peer, rw MsgReadWriter) error
+	Run func(peer *Peer, rw MsgReadWriter) error `json:"-"`
 
 	// NodeInfo is an optional helper method to retrieve protocol specific metadata
 	// about the host node.
-	NodeInfo func() interface{}
+	NodeInfo func() interface{} `json:"-"`
 
 	// PeerInfo is an optional helper method to retrieve protocol specific metadata
 	// about a certain peer in the network. If an info retrieval function is set,
 	// but returns nil, it is assumed that the protocol handshake is still running.
-	PeerInfo func(id enode.ID) interface{}
+	PeerInfo func(id enode.ID) interface{} `json:"-"`
 
 	// DialCandidates, if non-nil, is a way to tell Server about protocol-specific nodes
 	// that should be dialed. The server continuously reads nodes from the iterator and
 	// attempts to create connections to them.
-	DialCandidates enode.Iterator
+	DialCandidates enode.Iterator `json:"-"`
```

Comment:
What's the purpose of these?

Reply:
Through some path, exactly which I can look up, these get marshalled into json. Something holds a `Protocol` struct, and that something is then marshalled, causing these `func`s to also be marshalled. 
IIRC it was https://github.com/breml/errchkjson: 

>  Checks  types passed to the json encoding functions. Reports unsupported types  and optionally reports occations, where the check for the returned error  can be omitted.

Reply:
Ah, no I was wrong, it's `staticcheck` which doesn't like this: 

```
p2p/simulations/adapters/exec.go:182:32: SA1026: trying to marshal unsupported type func() interface{}, via x.Stack.P2P.Protocols[0].NodeInfo (staticcheck)
        confData, err := json.Marshal(confCopy)

```

Reply:
We can get around this by marking `p2p.Config.Protocols` as `json:"-"`.

Reply:
Yes. Is that preferred?

Reply:
Yes, because it avoids adding many `json:"-"` markers here, in `p2p.Protocol`, which is not meant to be marshaled at all.